fs-storage/src/file_storage.rs

```
@@ -0,0 +1,246 @@
+                    value_by_id.insert(id, value);
+                }
+
+                Ok(value_by_id)
+            }
+            Some(Err(e)) => Err(e.into()),
+            None => Err(ArklibError::Storage(
+                self.label.clone(),
+                "Storage file is missing header".to_owned(),
+            )),
+        }
+    }
+
+    /// Write data to file
+    ///
+    /// Data is a key-value mapping between [ResourceId] and a generic Value
+    pub fn write_file<K, V>(
+        &mut self,
+        value_by_id: &HashMap<K, V>,
+    ) -> Result<()>
+    where
+        K: Display,
+        V: Display,
+    {
+        fs::create_dir_all(self.path.parent().unwrap())?;
+        let file = File::create(&self.path)?;
+        let mut writer = BufWriter::new(file);
+
+        writer.write_all(
+            format!("{}{}\n", STORAGE_VERSION_PREFIX, STORAGE_VERSION)
+                .as_bytes(),
+        )?;
+
+        for (id, value) in value_by_id {
+            writer.write_all(
+                format!("{}{}{}\n", id, KEY_VALUE_SEPARATOR, value).as_bytes(),
+            )?;
+        }
+
+        let new_timestamp = fs::metadata(&self.path)?.modified()?;
+        if new_timestamp == self.timestamp {
+            return Err("Timestamp didn't update".into());
+        }
+        self.timestamp = new_timestamp;
+
+        log::info!(
+            "{} {} entries has been written",
+            self.log_prefix,
+            value_by_id.len()
+        );
+        Ok(())
+    }
+
+    /// Verify the version stored in the file header
+    fn verify_version(&self, header: &str) -> Result<()> {
+        if !header.starts_with(STORAGE_VERSION_PREFIX) {
+            return Err(ArklibError::Storage(
+                self.label.clone(),
+                "Unknown storage version".to_owned(),
+            ));
+        }
+
+        let version = header[STORAGE_VERSION_PREFIX.len()..]
+            .parse::<i32>()
+            .unwrap();
```

We can return the error here instead of using `unwrap()`